### PR TITLE
Add tree properties/transforms & reorganize

### PR DESCRIPTION
**Pull request by dcerkoney**
1. Add basic tree properties/transforms and graph operator traits.
2. Clean up comments/docstrings and reorganize (move tree properties to tree_properties.jl and transforms to transform.jl).
3. Add macro testset_skip to skip testsets and reorganize tests.
4. Remove restrictions on vertices in functions `linear_combination`, `replace_subgraph`, and `replace_subgraph!`.

### CODE REVIEW

File: src/computational_graph/tree_properties.jl [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+Base.IteratorEltype(::Type{<:TreeIterator{Graph{F,W}}}) where {F,W} = Base.HasEltype()
+Base.eltype(::Type{<:TreeIterator{Graph{F,W}}}) where {F,W} = Graph{F,W}
+
+function AbstractTrees.children(g::Graph)
+    return g.subgraphs
+end
+
+##################### Tree properties ########################### 
+
+# Does the graph have any children?
```
Comment (kunyuan) [inline src/computational_graph/tree_properties.jl:22]: are they external functions? If so, you need to add doc strings.
"""
      function ....

"""

File: test/computational_graph.jl [hunk 2]
```diff
@@ -108,20 +132,30 @@ end
     @test external_labels(g2) == [1, 4, 5, 3, 2]
 end
 
+@testset_skip "TODO: Update to new API" "Graph Operations" begin
```
Comment (kunyuan) [inline test/computational_graph.jl:135]: why you need new API for this test?

File: src/computational_graph/transform.jl [hunk 3]
```diff
@@ -87,21 +81,121 @@ end
...
+"""
+    function replace_subgraph!(g::Graph, w::Graph, m::graph)
 
-# function readDiag(io::IO)
+    In place function that replaces the children graph w in graph g with a new graph m.
+    Graphs w and m should have the same internal and external vertices, and topology
+"""
+function replace_subgraph!(g::Graph, w::Graph, m::Graph)
+    @assert !isleaf(g) "Target parent graph can not be a leaf"
+    @assert w.vertices == m.vertices "Old and new subgraph should have the same vertices"
```
Comment (kunyuan) [inline src/computational_graph/transform.jl:98]: 1. why replacing subgraph requires vertices to be the same? I think you can replace the subgraph as long as the external legs are the same?
2. It is also important that the Graph.order is the same

File: src/computational_graph/transform.jl [hunk 4]
```diff
@@ -87,21 +81,121 @@ end
...
+        g.subgraph_factors[1] *= child.subgraph_factors[1]
+        child.subgraph_factors[1] = 1
+    end
+    return g
+end
+
+"""
+    function merge_prefactors(g::Graph)
+   
+    Factorize the prefactors of a multiplicative graph g.
```
Comment (kunyuan) [inline src/computational_graph/transform.jl:172]: more detailed comments

File: src/computational_graph/transform.jl [hunk 5]
```diff
@@ -87,21 +81,121 @@ end
...
+        return eldest(g)
+    else
+        return g
+    end
+end
+
+"""
+    function inplace_prod(g::Graph)
+
+    Converts a unary Prod link to in-place form by propagating subgraph_factors up a level.
```
Comment (kunyuan) [inline src/computational_graph/transform.jl:154]: more comments, maybe with a simple example

File: src/computational_graph/transform.jl [hunk 6]
```diff
@@ -87,21 +81,121 @@ end
...
+end
 
-# function printDiag(diag::Union{Tuple,AbstractVector}, kwargs...)
+"""
+    function prune_trivial_unary(g::Graph)
 
-#     f = open("./DiagFiles/Diag.txt", "w")
-#     writedlm(f, [])
-# end
+    Returns a simplified copy of g if it represents a trivial unary operation.
```
Comment (kunyuan) [inline src/computational_graph/transform.jl:135]: give an example

File: src/computational_graph/transform.jl [hunk 7]
```diff
@@ -87,21 +81,121 @@ end
...
+"""
+    function replace_subgraph(g::Graph, w::Graph, m::graph)
 
-#     return Diagram
-# end
+    Generate a copy of graph g, with the children graph w replaced by a new graph m.
+    Graph w and m should have the same internal and external vertices, and topology
+"""
+function replace_subgraph(g::Graph, w::Graph, m::Graph)
+    @assert w.vertices == m.vertices "Old and new subgraph should have the same vertices"
```
Comment (kunyuan) [inline src/computational_graph/transform.jl:118]: same as before

File: src/computational_graph/tree_properties.jl (hunk 1 above)
Comment (dcerkoney) [reply src/computational_graph/tree_properties.jl:22]: I think we only need these helper functions internally, but I can add docstrings nonetheless.

File: src/computational_graph/transform.jl [hunk 8]
```diff
@@ -87,21 +81,121 @@ end
...
+        return g
+    end
+end
+
+"""
+    function inplace_prod(g::Graph)
+
+    Converts a unary Prod link to in-place form by propagating subgraph_factors up a level.
+"""
+function inplace_prod(g::Graph)
```
Comment (kunyuan) [inline src/computational_graph/transform.jl:156]: why inplace_prod only search for one level?

File: test/computational_graph.jl (hunk 2 above)
Comment (dcerkoney) [reply test/computational_graph.jl:135]: It uses the old interaction and feynman_diagram format. I will update it now.

File: src/computational_graph/transform.jl (hunk 8 above)
Comment (dcerkoney) [reply src/computational_graph/transform.jl:156]: `Base.*(g::Graph{F,W}, c::C)` already builds in a one-level inplace_prod (we should update that code to actually call the function). Hence, I think we can only form a Prod chain of length at most 2 during future transforms like expand, replace_subgraph, etc.

File: test/computational_graph.jl (hunk 2 above)
Comment (dcerkoney) [reply test/computational_graph.jl]: Fixed

File: src/computational_graph/transform.jl (hunk 8 above)
Comment (dcerkoney) [reply src/computational_graph/transform.jl]: Fixed

File: src/computational_graph/transform.jl (hunk 3 above)
Comment (dcerkoney) [reply src/computational_graph/transform.jl]: Resolved

File: src/computational_graph/tree_properties.jl (hunk 1 above)
Comment (dcerkoney) [reply src/computational_graph/tree_properties.jl]: Done. I also exported these functions for now; we can decide later if we want to move them to internal-only.

**Review by kunyuan — APPROVED** [review]

**Pull request merged**